File: server/src/main/java/rikka/shizuku/server/ShizukuClientManager.java
package rikka.shizuku.server;

import java.util.ArrayList;
import java.util.List;

import eu.darken.porter.core.CallerIdentity;
import eu.darken.porter.core.ClientCallback;

public class ShizukuClientManager extends ClientManager<ShizukuConfigManager> {
    private final List<ClientRecord> attached = new ArrayList<>();

    public ShizukuClientManager(ShizukuConfigManager configManager) {
        super(configManager);
    }

    @Override
    public synchronized ClientRecord attach(CallerIdentity identity, ClientCallback callback, String packageName, int apiVersion) {
        ClientRecord record = super.attach(identity, callback, packageName, apiVersion);
        if (record != null) {
            if (getConfigManager().isAccessPaused()) record.allowed = false;
            attached.removeIf(old -> findClient(old.uid, old.pid) != old);
            attached.add(record);
        }
        return record;
    }

    public synchronized List<ClientRecord> attachedClients() {
        attached.removeIf(record -> findClient(record.uid, record.pid) != record);
        return new ArrayList<>(attached);
    }
}

File: server/src/main/java/eu/darken/porter/privileged/ShizukuClientManager.kt
package eu.darken.porter.privileged

import eu.darken.porter.core.CallerIdentity
import eu.darken.porter.core.ClientCallback
import rikka.shizuku.server.ClientManager
import rikka.shizuku.server.ClientRecord

class ShizukuClientManager(configManager: ShizukuConfigManager) : ClientManager<ShizukuConfigManager>(configManager) {

    private val attached = ArrayList<ClientRecord>()

    @Synchronized
    override fun attach(identity: CallerIdentity, callback: ClientCallback, packageName: String, apiVersion: Int): ClientRecord? {
        val record = super.attach(identity, callback, packageName, apiVersion)
        if (record != null) {
            attached.removeIf { old -> findClient(old.uid, old.pid) !== old }
            attached.add(record)
        }
        return record
    }

    /**
     * A stored decision reaches a new process only while access is not paused, and only for the
     * installation it was made for.
     */
    override fun startsAllowed(identity: CallerIdentity, packageName: String): Boolean =
        !configManager.isAccessPaused &&
            super.startsAllowed(identity, packageName) &&
            configManager.verifiedForAttach(identity.uid, packageName)

    /** Told about every client whose process died; set once by the server that owns this manager. */
    @Volatile
    var onDeath: ((ClientRecord) -> Unit)? = null

    override fun onClientDied(record: ClientRecord) {
        onDeath?.invoke(record)
    }

    @Synchronized
    fun attachedClients(): List<ClientRecord> {
        attached.removeIf { record -> findClient(record.uid, record.pid) !== record }
        return ArrayList(attached)
    }
}
